### DISCUSSION TITLE
limits of orders-api

### DISCUSSION THREAD

**Discussion by vichilee** [amzn/selling-partner-api-models#3046] [General]
ordersV0.md 

getOrderBuyerInfo
getOrderAddress
getOrderItems
getOrderItemsBuyerInfo

<html>
<body>

Plan type | Rate (requests per second) | Burst
-- | -- | --
Default | 0.0055 | 20
Selling partner specific | Variable | Variable

</body>
</html>

If there are 1000 MFN orders, it means that getOrderAddress requests a maximum of 20 orders for the first time, and it takes 3 minutes to request the next order. It takes 2940 minutes to obtain the address information and product information of the 1000 orders. Is my understanding correct?

**Comment by jrl84** [comment]
Is this really meant to be the new restore rate of the orders endpoints from now on? Even getOrders and getOrder have this limit. One request per 3 minutes is breaking our app for our customers. This will force us to stick with the MWS API for the orders, this can't be the intended result for us and probably a lot of other users.

Could somebody from Amazon elaborate on this?

**Comment by michael-land** [comment]
180 second per order is breaking our app unless we stick with MWS :-(

**Comment by pierre-b** [comment]
Can Amazon explain why they put these restrictions please?
Either Amazon opens this endpoint fully or either they request an App review like PII... but at the moment it's frustrating.

**Comment by yberezkin** [comment]
Wait, hold on. The rate 0.0055 RPS | 20 Burst, where 3 min. come from?

**Comment by jrl84** [comment]
@yberezkin Its 0.0055 requests per second, so if you divide 1 by 0.0055 you get 180 seconds per request, which is 3 minutes. Well, you have 20 burst, but after these 20 you will have to wait 3 minutes for every new request.

**Comment by yberezkin** [comment]
@amz-tools Thanks, my fault.

**Comment by AdolphYu** [comment]
How to get historical orders

**Comment by fernlop** [comment]
The header for rate-limit is also present now, but it's not dynamic, fixed to 0.0055!

**Comment by xbaha** [comment]
has anyone tested if this rate limit is for real???

**Comment by fernlop** [comment]
@xbaha yes it really is..

**Comment by xbaha** [comment]
@fernlop on MWS it's 30/second on SPAPI it's 1/minute! 
and MWS will retire in September!
this would really break the app, and it becomes more complicated to deliver a useful app.

**Comment by fernlop** [comment]
@xbaha Did you read the rate limit header in the response? Does it say 0.016/s or why are you saying 1/minute?
I get 0.0055/s which is 1 every 3 minutes, but with 20 requests burst. 
It also breaks my app, my fix was to trigger the getOrders function every 3 minutes but that can't be the solution (If you have more than 1 order every 3 minutes you can't event get them all). 
They say its "dynamic" but they don't give any information how it works.

I don't understand how they think a "best practise" solution for interacting with this API could be. We are working with Lambda functions that get triggered in a specific amount of time because the is nothing like a notification for new orders. Should we work with reports? If yes the "new" REST-API doesn't make much sense if you must still do it the old way.

**Comment by xbaha** [comment]
@fernlop sorry, from my shock I wrote 1/m, but it's actually 3/minute which shocked me more...
This will break my app too, also getorderItems sometimes does not return the price, so I need to call it again!
it's either they add getOrdersItems to the API which take a list of orders, or increase the rate/second!

**Comment by kinjalnyc** [comment]
SP-API overall seems in terrible shape. We are having many issues so far.

**Reply by SergioNtw** [reply]
me too, bad response the app is unfunctional

**Comment by jhullparkseed** [comment]
Clearly I am having the same issues.  We use MWS to pull orders which contains all the information needed to push to our ERP.  I'm starting to look at SPAPI and hit the rate limits issue today when I tried a normal pull of orders.  Getting the orders is fine as that is one call, but now I need to make 3 extra calls, per order, to get all information.  This is much more inefficient even before the limits.

I tried an experiment with orders where I got more then 20 orders.  I then looped through orderItems and it threw a 429 at @ 22 order/items.   I then paused for 3 minutes (181 secs) ands continued.  I was able to get the next 22 and I repeated this till all items were pulled.

This is flat out unacceptable.  it took  over 6 minutes to get 63 items.

**Comment by xbaha** [comment]
it should take more for 63 orders, you get 20, then 1 every 3 minutes, that's (63-20)*3 = 129 minutes, how did you get them in 6 min???

**Comment by jhullparkseed** [comment]
This is a section of code that I used to test this (See below).  this is still not great and while on a daily basis our order level is manageable, we have peak seasons where ordering will exceed these limits.  The limits do not even match the example in one AWS documentation (something like 10000 requests per sec which makes no sense).

````
                foreach (Order order in resppnse.Payload.Orders)
                {
                    try
                    {
                        Amazon.SPAPI.Orders.Model.GetOrderItemsResponse ItemsResponse = ordersV0Api.GetOrderItems(order.AmazonOrderId);

   
```                     if (ordCnt > 4)
                        {
                            Thread.Sleep(1000);
                            ordCnt = 0;
                        }
                        ordCnt++;
                        foreach (OrderItem item in ItemsResponse.Payload.OrderItems)
                        {
                            Console.WriteLine(linecnt + "," + order.AmazonOrderId + "," + item.SellerSKU);
                            linecnt++;
                        }

                    }
                    catch (Exception ex)
                    {
                        Console.WriteLine(ex.Message);
                        Thread.Sleep(181000);
                        continue;
                    }

This is a run with time stamps:

Starting: 2021-06-04T14:29:48.3240895-04:00
1,xxx-xxxxx-xxxxx,100-xxx-PK
2,xxx-xxxxx-xxxxx,100-xxx-PK
3,xxx-xxxxx-xxxxx,100-xxx-PK
4,xxx-xxxxx-xxxxx,100-xxx-PK
5,xxx-xxxxx-xxxxx,100-xxx-PK
6,xxx-xxxxx-xxxxx,100-xxx-PK
7,xxx-xxxxx-xxxxx,100-xxx-PK
8,xxx-xxxxx-xxxxx,100-xxx-PK
9,xxx-xxxxx-xxxxx,100-xxx-PK
10,xxx-xxxxx-xxxxx,100-xxx-PK
11,xxx-xxxxx-xxxxx,100-xxx-PK
12,xxx-xxxxx-xxxxx,100-xxx-PK
13,xxx-xxxxx-xxxxx,100-xxx-PK
14,xxx-xxxxx-xxxxx,100-xxx-PK
15,xxx-xxxxx-xxxxx,100-xxx-PK
16,xxx-xxxxx-xxxxx,100-xxx-PK
17,xxx-xxxxx-xxxxx,100-xxx-PK
18,xxx-xxxxx-xxxxx,100-xxx-PK
19,xxx-xxxxx-xxxxx,100-xxx-PK
20,xxx-xxxxx-xxxxx,100-xxx-PK
21,xxx-xxxxx-xxxxx,100-xxx-PK
22,xxx-xxxxx-xxxxx,100-xxx-PK
Error calling GetOrderItems: {
  "errors": [
    {
      "message": "You exceeded your quota for the requested resource.",
     "code": "QuotaExceeded"
    }
  ]
}
pausing 181 secs: 6/4/2021 2:30:02 PM
23,xxx-xxxxx-xxxxx,100-xxx-PK
24,xxx-xxxxx-xxxxx,100-xxx-PK
25,xxx-xxxxx-xxxxx,100-xxx-PK
26,xxx-xxxxx-xxxxx,100-xxx-PK
27,xxx-xxxxx-xxxxx,100-xxx-PK
28,xxx-xxxxx-xxxxx,100-xxx-PK
29,xxx-xxxxx-xxxxx,100-xxx-PK
30,xxx-xxxxx-xxxxx,100-xxx-PK
31,xxx-xxxxx-xxxxx,100-xxx-PK
32,xxx-xxxxx-xxxxx,100-xxx-PK
33,xxx-xxxxx-xxxxx,100-xxx-PK
34,xxx-xxxxx-xxxxx,100-xxx-PK
35,xxx-xxxxx-xxxxx,100-xxx-PK
36,xxx-xxxxx-xxxxx,100-xxx-PK
37,xxx-xxxxx-xxxxx,100-xxx-PK
38,xxx-xxxxx-xxxxx,100-xxx-PK
39,xxx-xxxxx-xxxxx,100-xxx-PK
40,xxx-xxxxx-xxxxx,100-xxx-PK
41,xxx-xxxxx-xxxxx,100-xxx-PK
42,xxx-xxxxx-xxxxx,100-xxx-PK
43,xxx-xxxxx-xxxxx,100-xxx-PK
Error calling GetOrderItems: {
  "errors": [
    {
      "message": "You exceeded your quota for the requested resource.",
     "code": "QuotaExceeded"
    }
  ]
}
pausing 181 secs: 6/4/2021 2:33:16 PM
44,xxx-xxxxx-xxxxx,100-xxx-PK
45,xxx-xxxxx-xxxxx,100-xxx-PK
46,xxx-xxxxx-xxxxx,100-xxx-PK
47,xxx-xxxxx-xxxxx,100-xxx-PK
48,xxx-xxxxx-xxxxx,100-xxx-PK
49,xxx-xxxxx-xxxxx,100-xxx-PK
50,xxx-xxxxx-xxxxx,100-xxx-PK
51,xxx-xxxxx-xxxxx,100-xxx-PK
52,xxx-xxxxx-xxxxx,100-xxx-PK
53,xxx-xxxxx-xxxxx,100-xxx-PK
54,xxx-xxxxx-xxxxx,100-xxx-PK
55,xxx-xxxxx-xxxxx,100-xxx-PK
56,xxx-xxxxx-xxxxx,100-xxx-PK
57,xxx-xxxxx-xxxxx,100-xxx-PK
58,xxx-xxxxx-xxxxx,100-xxx-PK
59,xxx-xxxxx-xxxxx,100-xxx-PK
60,xxx-xxxxx-xxxxx,100-xxx-PK
61,xxx-xxxxx-xxxxx,100-xxx-PK
62,xxx-xxxxx-xxxxx,100-xxx-PK
ending: 2021-06-04T14:36:44.6264940-04:00

**Comment by xbaha** [comment]
Thank you, it seems that it's 20orders / 3 minutes, that more reasonable than what's in the docs....

**Comment by jhullparkseed** [comment]
To be specific, 20 OrderItem calls.  I have not yet tested this for the other two.  Even with 20 every 3 minutes....

If we average < 20 orders in each cycle (say 30 minutes) that is okay, but if our system goes down for x hours and we now have say 100 orders, for all three sub pulls that is:

100 / 20 = 5 pauses * 3 mins = 15 minutes of wait * each api call (buyer/address/items) that is 45 minutes of wait time or to put it another way, 45+minutes to pull down 100 orders.   That comes to 27 sec per order in that case!!

What would be better would be to create an API like 'GetAllOrderData" that combines the three other calls.  Even if the rate/burst is the same with a 3 min pause to get 20 more the number of calls is reduced and more potential orders are pulled (an less coding)

Hey Amazon Dev Team, build this in way before Sept 20th so it can be added to the tool box before shutting down MWS.

**Comment by bgkelly** [comment]
Is anyone at Amazon reading this thread?  The GetOrder* APIs are not fit for purpose.

**Comment by michael-land** [comment]
This already fixed in our case, you need to inspect the response header to get your real rate limit

**Comment by bgkelly** [comment]
> This already fixed in our case, you need to inspect the response header to get your real rate limit

What exactly is fixed?  The rate limits are far too low to be useful and the Developer Support guys don't seem to care.    The Reports API don't have a coherent set of reports that let you fetch arbitrary far back in time .... So if you're stuck on SP-API, then ones only choice is days of backfilling and retrying to fetch event a modestly sized FBA account's order history.

**Comment by michael-land** [comment]
Have you checked the rate limits in the response header? In our case, the rate now is 0.5 for `getOrderItem` and 0.04 for `getOrders`

> The x-amzn-RateLimit-Limit response header returns the usage plan rate limits that were applied to the requested operation. Rate limits for some selling partners will vary from the default rate and burst shown in the table above. For more information, see "Usage Plans and Rate Limits" in the Selling Partner API documentation.

**Comment by bgkelly** [comment]
I also see 0.5 for `getOrderItem` But still see rate limit errors even if I wait for 4 seconds between calls whereas 2 should be fine. The larger problem is if I can get 100 headers I should be able to fetch 100 items in the same call or with a call that has a matched limit.

**Comment by fr3shn3ss** [comment]
I ended up using "GET_FLAT_FILE_ACTIONABLE_ORDER_DATA_SHIPPING" report for unshipped orders. It takes less than a minute to return ~100 orders

**Reply by SergioNtw** [reply]
you need process the csv file, the file is well format and complete?

**Comment by Mike-the-one** [comment]
Sorry to bring up this old thread, last year Amazon stated this:

> Additionally by August 31, 2021, we will launch an update to the Selling Partner API for Orders to return the required data based on your eligibility within a single call similar to MWS Orders API.

What are the changes?

The SP-API (before Aug 31, 2021) to get orders so that merchant can fill requires the following calls:

```
getOrders
getOrderBuyerInfo
getOrderAddress
getOrderItems
getOrderItemsBuyerInfo
```

If I have 100 orders, I will have:

```
getOrders (x1)
getOrderBuyerInfo (x100)
getOrderAddress (x100)
getOrderItems (x100)
getOrderItemsBuyerInfo (x100)
```

So total 401 calls.

So, what exactly was changed after Aug 31/2021, will the `getOrders` call return all the information (buyer information, shipping address, order item details) for merchant to fulfill the order?

Thanks

**Reply by SergioNtw** [reply]
i don't found this information but is to much important, you find the endpoint that you comment?

**Comment by dbhandarkar** [comment]
> I ended up using "GET_FLAT_FILE_ACTIONABLE_ORDER_DATA_SHIPPING" report for unshipped orders. It takes less than a minute to return ~100 orders

Hello @fr3shn3ss , do you get all the PII data for FBM orders with this report type? I tried to use this, but unfortunately we do not get the BuyerName in the response

**Comment by fr3shn3ss** [comment]
I do see "recipient-name" and "ship-address-1". If you don't see them maybe that has something to do with your eligibility for PII

**Comment by dbhandarkar** [comment]
> I do see "recipient-name" and "ship-address-1". If you don't see them maybe that has something to do with your eligibility for PII

Hello @fr3shn3ss , Thank you for responding. But the BuyerName can be different from "recipient-name" or the "ship-address-1". I can place an order on behalf of m friend or as a gift. In this case, I will be the buyer and my friend is a receipeint. Do you see the same problem too?

**Comment by fr3shn3ss** [comment]
yeah this report doesn't seem to have it. You can use the getOrder API for that. I assume you don't need to pull this info in bulk unless if you have a unique use case. 

This report has the following columns
order-id	order-item-id	purchase-date	payments-date	reporting-date	promise-date	days-past-promise	buyer-email	buyer-phone-number	sku	product-name	quantity-purchased	quantity-shipped	quantity-to-ship	ship-service-level	recipient-name	ship-address-1	ship-address-2	ship-address-3	ship-city	ship-state	ship-postal-code	ship-country	is-business-order	purchase-order-number	price-designation

**Comment by lyoff** [comment]
this makes no sense. how do we pull all order details if seller has more than 480 orders per day???
How can they think of retiring MWS while there are such holes in SP API?

**Comment by fr3shn3ss** [comment]
> this makes no sense. how do we pull all order details if seller has more than 480 orders per day??? How can they think of retiring MWS while there are such holes in SP API?

Try this report ===> "GET_FLAT_FILE_ACTIONABLE_ORDER_DATA_SHIPPING"

**Comment by jhullparkseed** [comment]
"Try this report ===> "GET_FLAT_FILE_ACTIONABLE_ORDER_DATA_SHIPPING""

this is not a real solution.  Amazon has actual SP-API's functions to do this and they should better handle large order processing.

With that said, I built into my system an automatic throttle based on AMZ's calculations.  It counts the detail pulls and stops and waits the required 3 mins then starts again.  Our company averages close to 100 a day and growing so I'd love to not throttle all the time.

**Comment by Rajesh-ExertisGO** [comment]
We do it slightly different.
For getOrders API, we are passing  lastupdatedafter and lastupdatedbefore with RDT token get buyer info and PII data for all the orders in single call.
Once we get the orders, based on the status get the order lines.
Instead of running once a day we run it every 30 mins so we dont have to make too many getOrderItems api call at one time.

**Comment by adam-inco** [comment]
we have now switched to SP-API from MWS due to being told MWS is being stopped at the end of the month, these limits are madness and are not reasonable, once again amazon seem to ignore what sellers truly need

**Comment by ShaneOH** [comment]
Out of curiosity, what are you guys using the Orders API for that can't be accomplished by the Reports API if we're talking data imports (e.g. `GetOrders`)? The Reports API handles the volume issues you're facing and have been available from MWS too.

**Comment by bgkelly** [comment]
One example: I don't think the reportsAPI returns the OrderItemId ?

**Comment by napter** [comment]
Has Amazon weighed in on this? Has anyone reached out to find out why this limit is so laughably low?

**Comment by DawidWawrzyniak** [comment]
> One example: I don't think the reportsAPI returns the OrderItemId ?

This information in in report: GET_FLAT_FILE_ORDER_REPORT_DATA_INVOICING
If you do not see it in response, you need to check in seller central if you have enabled it to show in configuration report.

**Comment by xiaobai19950226** [comment]
Dear friends, our company has also encountered this issue now. There are over 800 orders that require getOrderAddress and getOrderBuyerInfo. Is there any way to solve this rate limit now? After all, even if we wait for one minute for each order, it's difficult for us to accept it

**Comment by grantrosse** [comment]
I was looking to pull historical order data and the solution is:
1. Find the right report type: https://developer-docs.amazon.com/sp-api/docs/report-type-values
2. Make post request for report type: https://developer-docs.amazon.com/sp-api/docs/reports-api-v2021-06-30-reference#post-reports2021-06-30reports
3. step 2 will yield a "reportId" in the reponse, use it to make a get request against this endpoint: https://developer-docs.amazon.com/sp-api/docs/reports-api-v2021-06-30-reference#get-reports2021-06-30reportsreportid
4. finally, step 3 will yield a "reportDocumentid" which you use in this endpoint: https://developer-docs.amazon.com/sp-api/docs/reports-api-v2021-06-30-reference#get-reports2021-06-30documentsreportdocumentid
5. this will then yield a 'url' to download the gzipped document

**Reply by SergioNtw** [reply]
Not is API call, the problem is always the bulk required as MWS API, now it to slow to offer good service to the seller and good service to amazon customer, because the quota system is damage include the health account. So, the real solution is quit the quota (this is not telephone service) y API rest revice, with auth user and permisions. This require atomic operation on the empoint to make eficiente service for the seller. Not relevant the orders that the seller have.
